### PR TITLE
Document automatic registration of extension compiler passes

### PR DESCRIPTION
**Pull request by wouterj**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes (https://github.com/symfony/symfony/pull/13761) |
| Applies to | 2.8+ |
| Fixed tickets | - |

### CODE REVIEW

File: components/dependency_injection/compilation.rst [hunk 1]
```diff
@@ -306,46 +306,85 @@ For more details, see :doc:`/cookbook/bundles/prepend_extension`, which
...
-being compiled::
+You can also execute custom code during compilation by writing your own
+compiler pass. By implementing
+:class:`Symfony\\Component\\DependencyInjection\\Compiler\\CompilerPassInterface`
+in your extension, the added ``process()`` method will be called during
+compilation::
 
+    // ...
     use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
-    use Symfony\Component\DependencyInjection\ContainerBuilder;
```
Comment (ogizanagi) [inline components/dependency_injection/compilation.rst]: This one is needed (used in `process`)

File: components/dependency_injection/compilation.rst [hunk 2]
```diff
@@ -306,46 +306,85 @@ For more details, see :doc:`/cookbook/bundles/prepend_extension`, which
...
 
-    class CustomCompilerPass implements CompilerPassInterface
+    class AcmeDemoExtension implements ExtensionInterface, CompilerPassInterface
```
Comment (ogizanagi) [inline components/dependency_injection/compilation.rst]: `AppExtension` ?

File: components/dependency_injection/compilation.rst (hunk 1 above)
Comment (WouterJ) [inline components/dependency_injection/compilation.rst]: Yeah, but we already added it in this file above and it's "included" in the `// ...` on line 320.

File: components/dependency_injection/compilation.rst (hunk 2 above)
Comment (WouterJ) [inline components/dependency_injection/compilation.rst]: This is the components section, so it does not have the convention of an AppBundle/AppExtension.

File: components/dependency_injection/compilation.rst [hunk 3]
```diff
@@ -306,46 +306,85 @@ For more details, see :doc:`/cookbook/bundles/prepend_extension`, which
 is specific to the Symfony Framework, but contains more details about this
 feature.
 
-Creating a Compiler Pass
-------------------------
+.. _components-di-compiler-pass:
```
Comment (xabbuh) [inline components/dependency_injection/compilation.rst]: We should also add a label for the old headline.

File: components/dependency_injection/compilation.rst [hunk 4]
```diff
@@ -306,46 +306,85 @@ For more details, see :doc:`/cookbook/bundles/prepend_extension`, which
...
-           // ...
+           // ... do something during the compilation
         }
+
+        // ...
     }
 
+.. versionadded:: 2.8
+    Prior to Symfony 2.8, extensions implementing ``CompilerPassInterface``
+    were not automatically registered. You need to register it as explained in
```
Comment (xabbuh) [inline components/dependency_injection/compilation.rst]: needed

Comment (xabbuh) [inline components/dependency_injection/compilation.rst]: [...] register them [...]

File: components/dependency_injection/compilation.rst [hunk 5]
```diff
@@ -306,46 +306,85 @@ For more details, see :doc:`/cookbook/bundles/prepend_extension`, which
...
+Execute Code During Compilation
+-------------------------------
 
-The compiler pass must have the ``process`` method which is passed the container
-being compiled::
+You can also execute custom code during compilation by writing your own
+compiler pass. By implementing
+:class:`Symfony\\Component\\DependencyInjection\\Compiler\\CompilerPassInterface`
+in your extension, the added ``process()`` method will be called during
+compilation::
```
Comment (xabbuh) [inline components/dependency_injection/compilation.rst]: We should tell  here that these extensions will be processed during the optimization step of the container compilation and that you cannot change that (which is actually different compared to using standalone compiler passes).

File: components/dependency_injection/compilation.rst [hunk 6]
```diff
@@ -306,46 +306,85 @@ For more details, see :doc:`/cookbook/bundles/prepend_extension`, which
...
 
-Registering a Compiler Pass
----------------------------
+Creating Separate Compiler Passes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-You need to register your custom pass with the container. Its process method
-will then be called when the container is compiled::
+Sometimes, you need to do more than one thing during compliation or want to use
+compiler passes without an extension. In this case, you can create a new class
```
Comment (xabbuh) [inline components/dependency_injection/compilation.rst]: Or you want to perform things in another steps of the compilation phase (see my comment above).

File: components/dependency_injection/tags.rst [hunk 7]
```diff
@@ -117,11 +117,13 @@ Notice that each was given a tag named ``acme_mailer.transport``. This is
 the custom tag that you'll use in your compiler pass. The compiler pass
 is what makes this tag "mean" something.
 
-Create a ``CompilerPass``
--------------------------
+.. _components-di-compiler-pass-tags:
 
-Your compiler pass can now ask the container for any services with the
-custom tag::
+Create a CompilerPass
```
Comment (xabbuh) [inline components/dependency_injection/tags.rst]: I would say let's use "Compiler Pass" (two words) here then.

File: components/dependency_injection/compilation.rst [hunk 8]
```diff
@@ -306,46 +306,85 @@ For more details, see :doc:`/cookbook/bundles/prepend_extension`, which
...
-methods described in the :doc:`/components/dependency_injection/definitions`.
-One common thing to do in a compiler pass is to search for all services
-that have a certain tag in order to process them in some way or dynamically
-plug each into some other service.
+methods described in :doc:`/components/dependency_injection/definitions`.
+
+.. note::
+
+    As a rule, only work with services definition in a compiler pass and do not
+    create service instances. Practically, this means using methods ``has()``,
```
Comment (javiereguiluz) [inline components/dependency_injection/compilation.rst]: `Practically` -> `In practice` ?

File: components/dependency_injection/compilation.rst (hunk 5 above)
Comment (WouterJ) [inline components/dependency_injection/compilation.rst]: Done.

**Comment by WouterJ** [comment]
Thanks for the detailed review guys, I've applied the comments.

File: components/dependency_injection/compilation.rst [hunk 9]
```diff
@@ -306,46 +306,94 @@ For more details, see :doc:`/cookbook/bundles/prepend_extension`, which
...
 The container's parameters and definitions can be manipulated using the
-methods described in the :doc:`/components/dependency_injection/definitions`.
-One common thing to do in a compiler pass is to search for all services
-that have a certain tag in order to process them in some way or dynamically
-plug each into some other service.
+methods described in :doc:`/components/dependency_injection/definitions`.
+
+.. note::
+
+    Please note that the ``process()`` method in the extension classes is
```
Comment (xabbuh) [inline components/dependency_injection/compilation.rst]: class

File: components/dependency_injection/compilation.rst [hunk 10]
```diff
@@ -306,46 +306,94 @@ For more details, see :doc:`/cookbook/bundles/prepend_extension`, which
...
+    called during the optimization step. You can read
+    :ref:`the next section <components-di-separate-compiler-passes>` if you
+    need to edit the container during another step.
+
+.. note::
+
+    As a rule, only work with services definition in a compiler pass and do not
+    create service instances. In practice, this means using methods ``has()``,
```
Comment (xabbuh) [inline components/dependency_injection/compilation.rst]: [...] using the methods [...]

**Comment by WouterJ** [comment]
Fixed

**Comment by xabbuh** [comment]
:+1:

**Pull request closed without merging**